Crash on gfx950 (MI355X) with fp8 KV: dim_quant 576 != 656 in dequantize_k_cache_paged

When a request is split across chunked-prefill steps (extend_prefix_lens != 0), the dense-MHA path reads the cached prefix KV via dequantize_k_cache_paged, which asserts the scaled fp8 layout (dim == 656). But on HIP (tilelang/aiter) the KV cache is the raw layout (dim == 576) → AssertionError: 576 != 656, scheduler aborts.

Repro: GLM-5.2-MXFP4 TP4, fp8 KV, --chunked-prefill-size 16384, random 1024/1024 conc ≥ 32 (or GSM8K high --parallel).

It looks like your run avoided the crash at conc 32/64 because **--chunked-prefill-size 131072** is larger than the batch's prefill size (1024 × 64 = 64K), so no request gets split.
However, the InferenceX benchmark may use conc 128/256 or higher, where the prefill size exceeds 131072 and requests get split again. So bumping --chunked-prefill-size higher isn't a safe or proper fix — it just pushes the threshold out. The underlying raw-vs-scaled KV layout mismatch should be handled directly instead.

--- a/python/sglang/srt/layers/attention/dsa_backend.py
+++ b/python/sglang/srt/layers/attention/dsa_backend.py
@@ -111,6 +111,20 @@ def _all_gather_dsa_trtllm_fp8_kv(
 
 _is_hip = is_hip()
 
+
+def _detect_gfx950() -> bool:
+    # gfx950 (MI355X) detection via gcnArchName; get_device_sm() is unreliable on ROCm.
+    if not _is_hip:
+        return False
+    try:
+        return "gfx950" in torch.cuda.get_device_properties(0).gcnArchName
+    except Exception:
+        return False
+
+
+# gfx950 supports the dense-MHA prefill fallback via aiter flash_attn_varlen_func.
+_is_gfx950 = _detect_gfx950()
+
 if _is_hip:
     from sglang.kernels.ops.attention.dsa.triton_kernel import get_valid_kv_indices
     from sglang.kernels.ops.quantization.fp8_kernel import fp8_dtype
@@ -2589,8 +2603,9 @@ class DeepseekSparseAttnBackend(
             f"cu_seqlens_k has {len(cu_seqlens_k)-1} requests"
         )
 
-        # Use TRTLLm ragged attention for SM100 (Blackwell/B200) to avoid FA4 accuracy issues
-        if self.device_sm_major >= 10:
+        # Use TRTLLm ragged attention for SM100 (Blackwell/B200) to avoid FA4 accuracy issues.
+        # ROCm (gfx950) always uses the aiter flash_attn_varlen_func path below.
+        if self.device_sm_major >= 10 and not _is_hip:
             import flashinfer
 
             seq_lens = metadata.cache_seqlens_int32
@@ -3022,11 +3037,13 @@ class DeepseekSparseAttnBackend(
             sum_seq_lens = sum(forward_batch.seq_lens_cpu)
             device_sm = get_device_sm()
 
-            # Requirements: H200/B200, short sequences, supported dtype, fits in chunk
+            # Requirements: H200/B200/MI355X, short sequences, supported dtype, fits in chunk
             self.use_mha = (
                 (
-                    device_sm == 90 or (device_sm >= 100 and device_sm < 110)
-                )  # SM90/SM100 only
+                    device_sm == 90
+                    or (device_sm >= 100 and device_sm < 110)
+                    or _is_gfx950
+                )  # SM90/SM100 (NVIDIA) or gfx950 (MI355X)
                 and max_kv_len
                 <= envs.SGLANG_DSA_PREFILL_DENSE_ATTN_KV_LEN_THRESHOLD.get()  # Short enough for MHA
                 and self.token_to_kv_pool.dtype in [torch.bfloat16, torch.float8_e4m3fn]

--- a/python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mha.py
+++ b/python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mha.py
@@ -589,6 +589,21 @@ class DeepseekMHAForwardMixin:
             kv_indices is not None
         ), "page_table_1_flattened should have been generated for FP8 MHA path"
 
+        if _use_aiter_gfx95:
+            # ROCm (gfx950) stores the FP8 MLA KV in the raw
+            # (kv_lora_rank + qk_rope_head_dim) layout, not the scaled 656-byte
+            # layout that dequantize_k_cache_paged expects (it asserts dim==656).
+            # Dequantize the raw layout via the pool's HIP-aware path instead —
+            # the same routine _get_mla_kv_buffer uses for the BF16 MHA path.
+            # Without this, a chunked-prefill split (extend_prefix_lens != 0) that
+            # reads cached prefix KV crashes with "576 != 656".
+            kv_indices = filter_dcp_local_kv_indices(kv_indices=kv_indices)
+            kv_a, k_pe = get_token_to_kv_pool().get_mla_kv_buffer(
+                self.attn_mha, kv_indices, torch.bfloat16
+            )
+            kv_a = kv_a.squeeze(1).contiguous()
+            return kv_a, k_pe
+
         kv_cache_fp8 = get_token_to_kv_pool().get_key_buffer(self.attn_mha.layer_id)
 
         kv_latent_bf16 = dequantize_k_cache_paged(kv_cache_fp8, kv_indices)

--- a/python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mha.py
+++ b/python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mha.py
@@ -297,10 +297,11 @@ class DeepseekMHAForwardMixin:
         q[..., self.qk_nope_head_dim :] = q_pe
 
         self._set_mla_kv_buffer(latent_cache, kv_a, k_pe, forward_batch)
-        if (
+        kv_refetched = (
             forward_batch.mha_one_shot
             and sum(forward_batch.extend_prefix_lens_cpu) != 0
-        ):
+        )
+        if kv_refetched:
             if (
                 self.use_dsa
                 and self.kv_cache_dtype == "fp8_e4m3"
@@ -345,9 +346,19 @@ class DeepseekMHAForwardMixin:
                 )
             )[0]
         else:
-            if _use_aiter_gfx95 and self.kv_b_proj.weight.dtype == torch.float8_e4m3fn:
+            if (
+                _use_aiter_gfx95
+                and self.kv_b_proj.weight.dtype == torch.float8_e4m3fn
+                and not kv_refetched
+            ):
                 kv = self.kv_b_proj(kv_a_quanted)[0]
             else:
+                # On a chunked-prefill split (kv_refetched), kv_a/k_pe were re-fetched as
+                # the FULL prefix+chunk sequence, but kv_a_quanted (from the fused
+                # norm+quant above) is only the current chunk — using it makes k_nope
+                # shorter than the full-length k_pe and crashes _concat_and_cast_mha_k
+                # (GLM-5.1-FP8, TP8: "expanded size 15497 must match 15753"). Fall back
+                # to the re-fetched bf16 kv_a; the fp8 kv_b_proj quantizes it inline.
                 kv = self.kv_b_proj(kv_a)[0]
             kv = kv.view(
                 -1, self.num_local_heads, self.qk_nope_head_dim + self.v_head_dim

